diff --git a/client/src/components/add-new-post/Editor.js b/client/src/components/add-new-post/Editor.js
index 50aaab5c..f2d00aee 100644
--- a/client/src/components/add-new-post/Editor.js
+++ b/client/src/components/add-new-post/Editor.js
@@ -6,7 +6,6 @@ import {
   Row,
   Col,
   FormCheckbox
-  
 } from "shards-react";
 import ReactQuill from "react-quill";
 import { Card, CardBody, Form, FormInput } from "shards-react";
@@ -14,6 +13,10 @@ import { Card, CardBody, Form, FormInput } from "shards-react";
 import "react-quill/dist/quill.snow.css";
 import "../../assets/quill.css";
 import {Redirect} from 'react-router-dom';
+import{
+  getFromStorage,
+  setInStorage
+} from "../../utils/storage.js";
 import axios from 'axios';
 
 
@@ -28,21 +31,58 @@ class Editor extends Component{
       error: '',
       saveData: false,
       isLoading: true,
+      isSubmitted: false,
     }
+    
   }
 
   componentDidMount(){
-    this.setState({isLoading: false});
+    const obj = getFromStorage('mainapp');
+    
+    if(obj && obj.token){
+      const { token } = obj;
+      fetch('api/user/verify?token=' + token)
+        .then(res => res.text())
+        .then(json => {
+          // console.log(json)
+          if(json.success){ 
+            this.setState({
+              token,
+              isLoading:false,
+              isSubmitted:false
+            });
+            
+            console.log(json);
+          }
+          else{
+            this.setState({
+              isLoading: false,
+              isSubmitted:false
+            });
+          }
+        }
+      );
+    }
+    else{
+      this.setState({
+        isLoading: false,
+        isSubmitted:false
+      });
+    }
   }
 
-  onSubmit(){
+  onSubmit(event){
+    event.preventDefault();
     const {
       title,
       description,
       sector,
-      amount
+      amount,
+      isSubmitted
     } = this.state;
 
+    console.log("Function started");
+
     fetch('http://localhost:5000/api/fundraiser/addFundraiser', {
       method: 'POST',
       headers:{
@@ -64,14 +104,16 @@ class Editor extends Component{
             title: '',
             description: '',
             sector: '',
-            amount: ''
+            amount: '',
+            isSubmitted: true
           }
         );
       }
       else{
         this.setState({
           addError: "Bhavika is crazy",
-          isLoading: false
+          isLoading: false,
+          isSubmitted: false
         });
       }
     })
@@ -81,64 +123,83 @@ class Editor extends Component{
     this.setState({[name]: event.target.value})
   }
 
+  handleChangeQuill(value){
+    // value = value.replace("<p>", "")
+    // value = value.replace("</p>", "")
+    console.log(value)
+    this.setState({description: value})
+  }
+
   render(){
     // if (this.state.saveData === true) {
     //   return <Redirect to='/blog-overview' />
     // }
-    return(
-      <Card small className="mb-3">
-        <CardBody>
-          <Form className="add-new-post">
-            <FormInput size="lg"  
-              className="mb-3" 
-              placeholder="Your Idea Title" 
-              name="title" 
-              defaultValue={this.state.title} 
-              onChange={this.handleChange('title')}/>
-            <ReactQuill 
-              placeholder="Describe your idea here. The funding recieved will depend heavily on how well the idea is described here." 
-              className="add-new-post__editor mb-1" 
-              name="description" defaultValue={this.state.description} />
-            <br></br>
-            <br></br>
-            <p>Select sector</p>
-              <FormCheckbox className="mb-1" name="sector" value={this.state.sector = 1} onChange={this.handleChange('sector')} >
-                Business
-              </FormCheckbox>
-              <FormCheckbox className="mb-1" name="sector" value={this.state.sector = 2} onChange={this.handleChange('sector')}>
-                Textile and Garments
-              </FormCheckbox>
-              <FormCheckbox className="mb-1" name="sector" value={this.state.sector = 3} onChange={this.handleChange('sector')} >
-                Tourism
-              </FormCheckbox>
-              <FormCheckbox className="mb-1" name="sector" value={this.state.sector = 4} onChange={this.handleChange('sector')} >
-                Other sectors
-              </FormCheckbox>
-
-              <br></br>
-              
-              <Row>
-                <Col md="2"><p large>Amount of funds to be raised</p></Col>
-                <Col md="6"><FormInput 
-                  size="lg" 
-                  placeholder="Enter the amount to be raised in Rupees" 
-                  name="amount" 
-                  defaultValue={this.state.amount} 
-                  onChange={this.handleChange('amount')} /></Col>
-              </Row>
+    const{
+      token,
+    } = this.state;
+    if(this.state.isSubmitted){
+      return <Redirect to='/blog-overview' />
+    }
+    else{
+    
+      return(
+        <Card small className="mb-3">
+          <CardBody>
+            <Form className="add-new-post">
+              <FormInput size="lg"  
+                className="mb-3" 
+                placeholder="Your Idea Title" 
+                name="title" 
+                defaultValue={this.state.title} 
+                onChange={this.handleChange('title')}/>
+              <ReactQuill 
+                placeholder="Describe your idea here. The funding recieved will depend heavily on how well the idea is described here." 
+                className="add-new-post__editor mb-1" 
+                name="description" value={this.state.description}
+                onChange={this.handleChangeQuill.bind(this)} />
               <br></br>
               <br></br>
-            <button
-              className="bg-primary text-white text-center rounded p-3 "
-              style={{ boxShadow: "inset 0 0 5px rgba(0,0,0,.2)" }}
-              onClick={this.onSubmit.bind(this)}>
-              Start your Fundraising Journey!
-            </button>
-          </Form>
-        </CardBody>
-      </Card>
-    );
-  }
-}
+              <p>Select sector</p>
+                <FormCheckbox className="mb-1" name="sector" value={this.state.sector = 1} onChange={this.handleChange('sector')} >
+                  Business
+                </FormCheckbox>
+                <FormCheckbox className="mb-1" name="sector" value={this.state.sector = 2} onChange={this.handleChange('sector')}>
+                  Textile and Garments
+                </FormCheckbox>
+                <FormCheckbox className="mb-1" name="sector" value={this.state.sector = 3} onChange={this.handleChange('sector')} >
+                  Tourism
+                </FormCheckbox>
+                <FormCheckbox className="mb-1" name="sector" value={this.state.sector = 4} onChange={this.handleChange('sector')} >
+                  Other sectors
+                </FormCheckbox>
+  
+                <br></br>
+                
+                <Row>
+                  <Col md="2"><p>Amount of funds to be raised</p></Col>
+                  <Col md="6"><FormInput 
+                    size="lg" 
+                    placeholder="Enter the amount to be raised in Rupees" 
+                    name="amount" 
+                    defaultValue={this.state.amount} 
+                    onChange={this.handleChange('amount')} /></Col>
+                </Row>
+                <br></br>
+                <br></br>
+              <button
+                className="bg-primary text-white text-center rounded p-3 "
+                style={{ boxShadow: "inset 0 0 5px rgba(0,0,0,.2)" }}
+                onClick={this.onSubmit.bind(this)}>
+                Start your Fundraising Journey!
+              </button>
+            </Form>
+          </CardBody>
+        </Card>
+      );
+    }
+    }
+    }
+    
+
 
 export default Editor;
diff --git a/client/src/components/layout/MainNavbar/NavbarNav/UserActions.js b/client/src/components/layout/MainNavbar/NavbarNav/UserActions.js
index 1abece29..1113d773 100644
--- a/client/src/components/layout/MainNavbar/NavbarNav/UserActions.js
+++ b/client/src/components/layout/MainNavbar/NavbarNav/UserActions.js
@@ -36,7 +36,7 @@ export default class UserActions extends React.Component {
             src={require("./../../../../images/avatars/1.svg")}
             alt="User Avatar"
           />{" "}
-          <span className="d-none d-md-inline-block">Bhavika Shahani</span>
+          <span className="d-none d-md-inline-block"></span>
         </DropdownToggle>
         <Collapse tag={DropdownMenu} right small open={this.state.visible}>
           <DropdownItem tag={Link} to="user-profile-lite">
diff --git a/client/src/components/login/LoginView.js b/client/src/components/login/LoginView.js
index 45a6f92b..62f76af2 100644
--- a/client/src/components/login/LoginView.js
+++ b/client/src/components/login/LoginView.js
@@ -75,6 +75,7 @@ class UserAccountDetails extends Component{
         isLoading: false
       });
     }
+    
     // onTextBoxChangeSignUpEmail = (event) => {
     //   this.setState({
     //     signUpEmail: event.target.value
@@ -160,7 +161,6 @@ class UserAccountDetails extends Component{
       return(<div><p>Loading...</p></div>);
     }
     if (isLoggedIn) {
-      // console.log(isLoggedIn)
       return <Redirect to='/blog-overview' />
     }
     // const {redirectToDashboard} = this.state
diff --git a/client/src/components/user-profile-lite/UserAccountDetails.js b/client/src/components/user-profile-lite/UserAccountDetails.js
index 8120c46e..e34544b6 100644
--- a/client/src/components/user-profile-lite/UserAccountDetails.js
+++ b/client/src/components/user-profile-lite/UserAccountDetails.js
@@ -29,7 +29,8 @@ class UserAccountDetails extends Component{
       isLoading: true,
       token: '',
       user_id: this.props.store.getUserId(),
-      user : ''
+      hits:[],
+      data: ''
     };
   }
   
@@ -43,22 +44,16 @@ class UserAccountDetails extends Component{
           'Content-Type': 'application/json'
         },
       })
-      // .then(res => res.json()
-      .then(res => {
-        this.setState({user : res.json()})
-        // console.log("user is" + this.state.user);
-        return this.state.user;
-      })
+      // .then(res => res.json())
+      .then(data => this.setState({ hits: data.hits }));
+      console.log("hits" + this.state.hits)
+      console.log("data" + this.state.data)
   }
   
   render(){
     // (jsons.length > 0) ? jsons.map( (json) => {
-        const user = this.state.user;
-        var users;
-
-        // console.log(user);
-        
-          return(   
+      const {data, hits} = this.state;
+          return(
             <Card small className="mb-4"> 
               <CardHeader className="border-bottom">
                 {/*<h6 className="m-0">{this.props.title}</h6>*/}
@@ -66,133 +61,136 @@ class UserAccountDetails extends Component{
               <ListGroup>
                 
                 <ListGroupItem className="p-3">
-                  {/*{console.log(this.state.user)}*/}
-                {users = Object.keys(this.state.user).forEach(function(element) {
-           
                   <Row>
                     <Col>
                       <Form>
                         <Row form>
-
-                          
                           {/* First Name */}
-                          
+                          {data.hits.map((hit) =>
                           <Col md="6" className="form-group">
                             
                             <label htmlFor="feFirstName">First Name</label> 
                             <FormInput disabled
                               id="feFirstName"
                               placeholder="First Name"
-                              value={this.state.user.firstname}
+                              value={hit.firstname}
                               onChange={() => {}}
                             />
                           </Col>
+                          )}
                           
                           
                           {/* Last Name */}
+                          {data.hits.map((hit) =>
                           <Col md="6" className="form-group">
                             <label htmlFor="feLastName">Last Name</label>
                             <FormInput disabled
                               id="feLastName"
                               placeholder="Last Name"
-                              value={this.state.user.lastname}
+                              value={hit.lastname}
                               onChange={() => {}}
                             />
                           </Col>
+                          )}
                         </Row>
                         <Row form>
                           {/* Email */}
-                          
+                          {data.hits.map((hit) =>
                           <Col md="6" className="form-group">
                             <label htmlFor="feEmail">Email</label>
                             <FormInput disabled
                               type="email"
                               id="feEmail"
                               placeholder="Email Address"
-                              value={this.state.user.email_id}
+                              value={hit.email_id}
                               onChange={() => {}}
                               autoComplete="email"
                             />
                           </Col>
+                          )}
                         </Row>
                         {/*Address*/}
                         <Row form>
-                          
+                          {data.hits.map((hit) =>
                         <FormGroup>
                           <label htmlFor="feAddress">Address</label>
                           <FormInput disabled
                             id="feAddress"
                             placeholder="Address"
-                            value={this.state.user.address}
+                            value={hit.address}
                             onChange={() => {}}
                           />
                           
                         </FormGroup>
- 
+                        )}
                         </Row>
                         <Row form>
                           {/* City */}
+                          {data.hits.map((hit) => 
                           <Col md="6" className="form-group">
                             <label htmlFor="feCity">City</label>
                             <FormInput disabled
                               id="feCity"
                               placeholder="City"
-                              value = {this.state.user.city}
+                              value = {hit.city}
                               onChange={() => {}}
                             />
                           </Col>
-                          {/* State */} 
+                          )}
+                          {/* State */}
+                          {data.hits.map((hit) => 
                           <Col md="4" className="form-group">
                             <label htmlFor="feInputState">State</label>
                             <FormInput disabled
                               id="feState"
                               placeholder="State"
-                              value = {this.state.user.state}
+                              value = {hit.state}
                               onChange={() => {}}
                             />
                           </Col>
+                          )}
                           {/*country*/}
-                           
+                          {data.hits.map((hit) => 
                           <Col md="4" className="form-group">
                             <label htmlFor="feInputCountry">State</label>
                             <FormInput disabled
                               id="feCountry"
                               placeholder="Country"
-                              value = {this.state.user.country}
+                              value = {hit.country}
                               onChange={() => {}}
                             />
                           </Col>
+                          )}
                           {/* Zip Code */}
-                           
+                          {data.hits.map((hit) => 
                           <Col md="2" className="form-group">
                             <label htmlFor="feZipCode">Zip</label>
                             <FormInput disabled
                               id="feZipCode"
                               placeholder="Zip"
-                              value={this.state.user.pincode}
+                              value={hit.pincode}
                               onChange={() => {}}
                             />
                           </Col>
-                          
+                          )}
                         </Row>
                         <Row form>
                           
                           {/* Description */}
+                          {data.hits.map((hit) => 
                           <Col md="12" className="form-group">
                             <label htmlFor="feDescription">Description</label>
-                            <FormTextarea disabled id="feDescription" value= {this.state.user.description} rows="5" />
+                            <FormTextarea disabled id="feDescription" value= {hit.description} rows="5" />
                           </Col>
-                          
+                          )}
                         </Row>
                       </Form>
                     </Col>
                   </Row>
-                  }, this)}
                 </ListGroupItem>
+                )}
               </ListGroup>
-              
             </Card>
-       
           );
   }
 }
diff --git a/client/src/views/BlogOverview.js b/client/src/views/BlogOverview.js
index 18028bd3..b1199d30 100644
--- a/client/src/views/BlogOverview.js
+++ b/client/src/views/BlogOverview.js
@@ -17,96 +17,278 @@ import UsersByDevice from "./../components/blog/UsersByDevice";
 import Discussions from "./../components/blog/Discussions";
 import TopReferrals from "./../components/common/TopReferrals";
 
+import{
+  getFromStorage,
+  setInStorage
+} from "./../utils/storage";
 
-const BlogOverview = ({ smallStats, PostsListOne }) => (
-  <Container fluid className="main-content-container px-4">
-    {/* Page Header */}
-    <Row noGutters className="page-header py-4">
-      <PageTitle title="Dashboard" subtitle="Profile Overview" className="text-sm-left mb-3" />
-    </Row>
+class BlogOverview extends React.Component {
+  constructor(props){
+    super(props)
+    console.log(props)
+    this.state={
+      smallStats:  [
+        {
+          label: "Page Hits",
+          value: "2,390",
+          percentage: "4.7%",
+          increase: true,
+          chartLabels: [null, null, null, null, null, null, null],
+          attrs: { md: "6", sm: "6" },
+          datasets: [
+            {
+              label: "Today",
+              fill: "start",
+              borderWidth: 1.5,
+              backgroundColor: "rgba(0, 184, 216, 0.1)",
+              borderColor: "rgb(0, 184, 216)",
+              data: [1, 2, 1, 3, 5, 4, 7]
+            }
+          ]
+        },
+        {
+          label: "Ideas",
+          value: "182",
+          percentage: "12.4",
+          increase: true,
+          chartLabels: [null, null, null, null, null, null, null],
+          attrs: { md: "6", sm: "6" },
+          datasets: [
+            {
+              label: "Today",
+              fill: "start",
+              borderWidth: 1.5,
+              backgroundColor: "rgba(23,198,113,0.1)",
+              borderColor: "rgb(23,198,113)",
+              data: [1, 2, 3, 3, 3, 4, 4]
+            }
+          ]
+        },
+        {
+          label: "Registrations",
+          value: "8,147",
+          percentage: "3.8%",
+          increase: false,
+          decrease: true,
+          chartLabels: [null, null, null, null, null, null, null],
+          attrs: { md: "4", sm: "6" },
+          datasets: [
+            {
+              label: "Today",
+              fill: "start",
+              borderWidth: 1.5,
+              backgroundColor: "rgba(255,180,0,0.1)",
+              borderColor: "rgb(255,180,0)",
+              data: [2, 3, 3, 3, 4, 3, 3]
+            }
+          ]
+        },
+        {
+          label: "New Users",
+          value: "29",
+          percentage: "2.71%",
+          increase: false,
+          decrease: true,
+          chartLabels: [null, null, null, null, null, null, null],
+          attrs: { md: "4", sm: "6" },
+          datasets: [
+            {
+              label: "Today",
+              fill: "start",
+              borderWidth: 1.5,
+              backgroundColor: "rgba(255,65,105,0.1)",
+              borderColor: "rgb(255,65,105)",
+              data: [1, 7, 1, 3, 1, 4, 8]
+            }
+          ]
+        },
+        {
+          label: "Contributions",
+          value: "17,281",
+          percentage: "2.4%",
+          increase: false,
+          decrease: true,
+          chartLabels: [null, null, null, null, null, null, null],
+          attrs: { md: "4", sm: "6" },
+          datasets: [
+            {
+              label: "Today",
+              fill: "start",
+              borderWidth: 1.5,
+              backgroundColor: "rgb(0,123,255,0.1)",
+              borderColor: "rgb(0,123,255)",
+              data: [3, 2, 3, 2, 4, 5, 4]
+            }
+          ]
+        }
+      ],
+      PostsListOne: [
+        {
+          backgroundImage: require("../images/Logo3.png"),
+          category: "Aviation",
+          categoryTheme: "dark",
+          author: "Abdul Rashid ",
+          authorAvatar: require("../images/avatars/1.jpg"),
+          title: "Conduct at an replied removal an amongst",
+          body:
+            "However venture pursuit he am mr cordial. Forming musical am hearing studied be luckily. But in for determine what would see...",
+          date: "28 February 2019"
+        },
+        {
+          backgroundImage: require("../images/Logo5.png"),
+          category: "Automobile",
+          categoryTheme: "info",
+          author: "Hassan Izz-Al-Din",
+          authorAvatar: require("../images/avatars/2.jpg"),
+          title: "Off tears are day blind smile alone had ready",
+          body:
+            "Is at purse tried jokes china ready decay an. Small its shy way had woody downs power. To denoting admitted speaking learning my...",
+          date: "29 February 2019"
+        },
+        {
+          backgroundImage: require("../images/Logo3.png"),
+          category: "Chemicals",
+          categoryTheme: "royal-blue",
+          author: "Osama bin Laden",
+          authorAvatar: require("../images/avatars/2.jpg"),
+          title: "Difficult in delivered extensive at direction",
+          body:
+            "Is at purse tried jokes china ready decay an. Small its shy way had woody downs power. To denoting admitted speaking learning my...",
+          date: "29 February 2019"
+        },
+        {
+          backgroundImage: require("../images/Logo5.png"),
+          category:"Defence",
+          categoryTheme: "warning",
+          author: "Khalid Sheikh Mohammed",
+          authorAvatar: require("../images/avatars/3.jpg"),
+          title: "It so numerous if he may outlived disposal",
+          body:
+            "How but sons mrs lady when. Her especially are unpleasant out alteration continuing unreserved ready road market resolution...",
+          date: "29 February 2019"
+        }
+      ]
+    }
 
-    <Row>
-      {PostsListOne.map((post, idx) => (
-        <Col lg="3" md="6" sm="12" className="mb-4" key={idx}>
-          <Card small className="card-post card-post--1">
-            <div
-              className="card-post__image"
-              style={{ backgroundImage: `url(${post.backgroundImage})` }}
-            >
-              <Badge
-                pill
-                className={`card-post__category bg-${post.categoryTheme}`}
-              >
-                {post.category}
-              </Badge>
-              <div className="card-post__author d-flex">
-                <a
-                  href="/"
-                  className="card-post__author-avatar card-post__author-avatar--small"
-                  style={{ backgroundImage: `url('${post.authorAvatar}')` }}
+  }
+
+  componentDidMount() {
+    fetch('http://localhost:5000/api/fundraiser/allFundraisers', {
+      method: 'GET',
+      headers:{
+        'Content-Type': 'application/json'
+      }
+    })
+    .then(res => res.json())
+    .then(json => {
+      console.log('json', json);
+      if(json){
+        this.setState({
+          PostsListOne: json.docs
+        });
+      }
+      else{
+        this.setState({
+          addError: "Bhavika is crazy",
+          isLoading: false,
+          isSubmitted: false
+        });
+      }
+    })
+  }
+
+  render(){
+    return(
+      <Container fluid className="main-content-container px-4">
+        {/* Page Header */}
+        <Row noGutters className="page-header py-4">
+          <PageTitle title="Dashboard" subtitle="Profile Overview" className="text-sm-left mb-3" />
+        </Row>
+
+        <Row>
+          {this.state.PostsListOne.map((post, idx) => (
+            <Col lg="3" md="6" sm="12" className="mb-4" key={idx}>
+              <Card small className="card-post card-post--1">
+                <div
+                  className="card-post__image"
+                  style={{ backgroundImage: `url(${post.backgroundImage})` }}
                 >
-                  Written by {post.author}
-                </a>
-              </div>
-            </div>
-            <CardBody>
-              <h5 className="card-title">
-                <a href="/" className="text-fiord-blue">
-                  {post.title}
-                </a>
-              </h5>
-              <p className="card-text d-inline-block mb-3">{post.body}</p>
-              <span className="text-muted">{post.date}</span>
-            </CardBody>
-          </Card>
-        </Col>
-      ))}
-    </Row>
+                  <Badge
+                    pill
+                    className={`card-post__category bg-${post.categoryTheme}`}
+                  >
+                    {post.category}
+                  </Badge>
+                  <div className="card-post__author d-flex">
+                    <a
+                      href="/"
+                      className="card-post__author-avatar card-post__author-avatar--small"
+                      style={{ backgroundImage: `url('${post.authorAvatar}')` }}
+                    >
+                      Written by {post.author}
+                    </a>
+                  </div>
+                </div>
+                <CardBody>
+                  <h5 className="card-title">
+                    <a href="/" className="text-fiord-blue">
+                      {post.title}
+                    </a>
+                  </h5>
+                  <p className="card-text d-inline-block mb-3">{post.body}</p>
+                  <span className="text-muted">{post.date}</span>
+                </CardBody>
+              </Card>
+            </Col>
+          ))}
+        </Row>
 
-    <Row>
-      {/* Discussions */}
-      <Col lg="8" md="12" sm="12" className="mb-4">
-        <Discussions />
-      </Col>
+        <Row>
+          {/* Discussions */}
+          <Col lg="8" md="12" sm="12" className="mb-4">
+            <Discussions />
+          </Col>
 
-      {/* Top Referrals */}
-      <Col lg="4" md="12" sm="12" className="mb-4">
-        <TopReferrals />
-      </Col>
+          {/* Top Referrals */}
+          <Col lg="4" md="12" sm="12" className="mb-4">
+            <TopReferrals />
+          </Col>
 
-      {/* Users Overview */}
-      <Col lg="8" md="12" sm="12" className="mb-4">
-        <UsersOverview />
-      </Col>
+          {/* Users Overview */}
+          <Col lg="8" md="12" sm="12" className="mb-4">
+            <UsersOverview />
+          </Col>
 
-      {/* Users by Device */}
-      <Col lg="4" md="6" sm="12" className="mb-4">
-        <UsersByDevice />
-      </Col>
-    </Row>
+          {/* Users by Device */}
+          <Col lg="4" md="6" sm="12" className="mb-4">
+            <UsersByDevice />
+          </Col>
+        </Row>
 
-    {/* Small Stats Blocks */}
-    <Row>
-      {smallStats.map((stats, idx) => (
-        <Col className="col-lg mb-4" key={idx} {...stats.attrs}>
-          <SmallStats
-            id={`small-stats-${idx}`}
-            variation="1"
-            chartData={stats.datasets}
-            chartLabels={stats.chartLabels}
-            label={stats.label}
-            value={stats.value}
-            percentage={stats.percentage}
-            increase={stats.increase}
-            decrease={stats.decrease}
-          />
-        </Col>
-      ))}
-    </Row>
+        {/* Small Stats Blocks */}
+        <Row>
+          {this.state.smallStats.map((stats, idx) => (
+            <Col className="col-lg mb-4" key={idx} {...stats.attrs}>
+              <SmallStats
+                id={`small-stats-${idx}`}
+                variation="1"
+                chartData={stats.datasets}
+                chartLabels={stats.chartLabels}
+                label={stats.label}
+                value={stats.value}
+                percentage={stats.percentage}
+                increase={stats.increase}
+                decrease={stats.decrease}
+              />
+            </Col>
+          ))}
+        </Row>
 
 
-  </Container>
-);
+      </Container>
+    )
+  }
+}
 
 BlogOverview.propTypes = {
   /**
@@ -117,147 +299,7 @@ BlogOverview.propTypes = {
 };
 
 BlogOverview.defaultProps = {
-  smallStats: [
-    {
-      label: "Page Hits",
-      value: "2,390",
-      percentage: "4.7%",
-      increase: true,
-      chartLabels: [null, null, null, null, null, null, null],
-      attrs: { md: "6", sm: "6" },
-      datasets: [
-        {
-          label: "Today",
-          fill: "start",
-          borderWidth: 1.5,
-          backgroundColor: "rgba(0, 184, 216, 0.1)",
-          borderColor: "rgb(0, 184, 216)",
-          data: [1, 2, 1, 3, 5, 4, 7]
-        }
-      ]
-    },
-    {
-      label: "Ideas",
-      value: "182",
-      percentage: "12.4",
-      increase: true,
-      chartLabels: [null, null, null, null, null, null, null],
-      attrs: { md: "6", sm: "6" },
-      datasets: [
-        {
-          label: "Today",
-          fill: "start",
-          borderWidth: 1.5,
-          backgroundColor: "rgba(23,198,113,0.1)",
-          borderColor: "rgb(23,198,113)",
-          data: [1, 2, 3, 3, 3, 4, 4]
-        }
-      ]
-    },
-    {
-      label: "Registrations",
-      value: "8,147",
-      percentage: "3.8%",
-      increase: false,
-      decrease: true,
-      chartLabels: [null, null, null, null, null, null, null],
-      attrs: { md: "4", sm: "6" },
-      datasets: [
-        {
-          label: "Today",
-          fill: "start",
-          borderWidth: 1.5,
-          backgroundColor: "rgba(255,180,0,0.1)",
-          borderColor: "rgb(255,180,0)",
-          data: [2, 3, 3, 3, 4, 3, 3]
-        }
-      ]
-    },
-    {
-      label: "New Users",
-      value: "29",
-      percentage: "2.71%",
-      increase: false,
-      decrease: true,
-      chartLabels: [null, null, null, null, null, null, null],
-      attrs: { md: "4", sm: "6" },
-      datasets: [
-        {
-          label: "Today",
-          fill: "start",
-          borderWidth: 1.5,
-          backgroundColor: "rgba(255,65,105,0.1)",
-          borderColor: "rgb(255,65,105)",
-          data: [1, 7, 1, 3, 1, 4, 8]
-        }
-      ]
-    },
-    {
-      label: "Contributions",
-      value: "17,281",
-      percentage: "2.4%",
-      increase: false,
-      decrease: true,
-      chartLabels: [null, null, null, null, null, null, null],
-      attrs: { md: "4", sm: "6" },
-      datasets: [
-        {
-          label: "Today",
-          fill: "start",
-          borderWidth: 1.5,
-          backgroundColor: "rgb(0,123,255,0.1)",
-          borderColor: "rgb(0,123,255)",
-          data: [3, 2, 3, 2, 4, 5, 4]
-        }
-      ]
-    }
-  ],
-  PostsListOne: [
-    {
-      backgroundImage: require("../images/Logo3.png"),
-      category: "Aviation",
-      categoryTheme: "dark",
-      author: "Abdul Rashid ",
-      authorAvatar: require("../images/avatars/1.jpg"),
-      title: "Conduct at an replied removal an amongst",
-      body:
-        "However venture pursuit he am mr cordial. Forming musical am hearing studied be luckily. But in for determine what would see...",
-      date: "28 February 2019"
-    },
-    {
-      backgroundImage: require("../images/Logo5.png"),
-      category: "Automobile",
-      categoryTheme: "info",
-      author: "Hassan Izz-Al-Din",
-      authorAvatar: require("../images/avatars/2.jpg"),
-      title: "Off tears are day blind smile alone had ready",
-      body:
-        "Is at purse tried jokes china ready decay an. Small its shy way had woody downs power. To denoting admitted speaking learning my...",
-      date: "29 February 2019"
-    },
-    {
-      backgroundImage: require("../images/Logo3.png"),
-      category: "Chemicals",
-      categoryTheme: "royal-blue",
-      author: "Osama bin Laden",
-      authorAvatar: require("../images/avatars/2.jpg"),
-      title: "Difficult in delivered extensive at direction",
-      body:
-        "Is at purse tried jokes china ready decay an. Small its shy way had woody downs power. To denoting admitted speaking learning my...",
-      date: "29 February 2019"
-    },
-    {
-      backgroundImage: require("../images/Logo5.png"),
-      category:"Defence",
-      categoryTheme: "warning",
-      author: "Khalid Sheikh Mohammed",
-      authorAvatar: require("../images/avatars/3.jpg"),
-      title: "It so numerous if he may outlived disposal",
-      body:
-        "How but sons mrs lady when. Her especially are unpleasant out alteration continuing unreserved ready road market resolution...",
-      date: "29 February 2019"
-    }
-  ]
+  
 };
 
 export default BlogOverview;
diff --git a/node_modules/bcrypt/lib/binding/bcrypt_lib.node b/node_modules/bcrypt/lib/binding/bcrypt_lib.node
index 8ab35b98..8981b0a5 100644
Binary files a/node_modules/bcrypt/lib/binding/bcrypt_lib.node and b/node_modules/bcrypt/lib/binding/bcrypt_lib.node differ
diff --git a/routes/api/fundraiser.js b/routes/api/fundraiser.js
index 0fd24f7e..c02bf9a6 100644
--- a/routes/api/fundraiser.js
+++ b/routes/api/fundraiser.js
@@ -61,14 +61,14 @@ function addFundraiser (req,res, next) {
                     message: 'Added Data.'
                 });
             }
-    });
+        })
     // .then(fundraiser => console.log(res.json(fundraiser)));
 };
 router.get('/allFundraisers', getAllFundraisers);
 
 function getAllFundraisers(req, res, next){
     Fundraiser.find({}, 
-        err => {
+        (err, docs) => {
             if(err) {
                 return res.send(
                     {
@@ -77,12 +77,15 @@ function getAllFundraisers(req, res, next){
                     }
                 )
             }
+            return res.send(
+            {
+                success: true,
+                message:'Finally working',
+                docs
+            }
+            );
         });
 
-        return res.send({
-            success: true,
-            message:'Finally working'
-        });
 }
  
 module.exports = router;
\ No newline at end of file
